### PR TITLE
[PWA-814] Enable shoppers to change locale using a store view switcher

### PR DESCRIPTION
**Pull request by eug123**
## Description

Adding store switcher component to allow visitor to switch store views and get content translated to respective store view locale.

## Related Issue

https://jira.corp.magento.com/browse/PWA-814

### Verification Stakeholders

@jimbo
@sirugh 
@dpatil-magento 
### Verification Steps

1. Configure Magento to have more then one store view
2. See store switcher appeared in the PWA header for desktop / Side navigation for mobile screen
3. Use switcher to change store view/locale
4. Verify that data been retrieved according to selected store view

## Checklist

- [ ] I have added tests to cover my changes, if necessary.
* I have updated the documentation accordingly, if necessary.

### CODE REVIEW

File: packages/peregrine/lib/talons/Header/useStoreSwitcher.js
Comment (sirugh) [inline packages/peregrine/lib/talons/Header/useStoreSwitcher.js:65]: I would replace this with a `writeQuery`

File: packages/venia-concept/src/index.js
Comment (sirugh) [inline packages/venia-concept/src/index.js:42]: Remove the default state above, preferably.

```suggestion
    const storeCode = storage.getItem('store_view').code || STORE_VIEW_CODE;
```

File: packages/venia-ui/lib/components/MagentoRoute/magentoRoute.js
Comment (sirugh) [inline packages/venia-ui/lib/components/MagentoRoute/magentoRoute.js:25]: `useQuery(getStoreViewCode)`

File: packages/venia-ui/lib/components/Header/storeSwitcher.js
Comment (sirugh) [inline packages/venia-ui/lib/components/Header/storeSwitcher.js:100]: There's gotta be a better way. Have you run the implementation by UX first and foremost? If this UX is what they want, then that means we need some way to conditionally use either `5rem` or `7.5rem`. Feels odd to do this in the component though.

Also, if we're going to keep this, we should use an effect, on mount, so that it does not continue to call the setter every render.

File: packages/venia-ui/lib/components/MagentoRoute/magentoRoute.js
Comment (sirugh) [inline packages/venia-ui/lib/components/MagentoRoute/magentoRoute.js:10]: Third party deps should be near the top of the imports.

Can you make sure this change is applied throughout the changeset?

File: packages/venia-ui/lib/components/MagentoRoute/magentoRoute.js
Comment (sirugh) [inline packages/venia-ui/lib/components/MagentoRoute/magentoRoute.js:38]: This will work because `formatMessage` falls back to the `id` if there is no defined message in the json. That said, I think this should actually be something like:

```js
const errorMessage = formatMessage({ 
  id: routeError,
  defaultMessage: MESSAGES.get(routeError);
});

...
<h1>
  {errorMessage}
</h1>
```

And you'd define the mapping of route error within in the translation json. You'd probably need to modify the keys to be something like `magentoRoute.notFoundError`. etc.

File: packages/venia-ui/lib/components/MagentoRoute/magentoRoute.js
Comment (sirugh) [inline packages/venia-ui/lib/components/MagentoRoute/magentoRoute.js:45]: Same idea here. We can use the map strings as a default message but should also make sure they are in the json.

File: packages/venia-ui/lib/components/Navigation/navigation.js
Comment (sirugh) [inline packages/venia-ui/lib/components/Navigation/navigation.js:72]: Usually we indicate mobile using media queries in css. I think it makes more sense to do it that way anyways. The component doesn't seem too complex, just a horizontal bar with an option/select.

File: packages/venia-ui/lib/components/Header/header.css
Comment (sirugh) [inline packages/venia-ui/lib/components/Header/header.css:2]: Do we define other global variables in such a way? I'd like to talk about this in dev time.

cc @jimbo

File: packages/venia-ui/lib/components/SearchPage/searchPage.js
Comment (sirugh) [inline packages/venia-ui/lib/components/SearchPage/searchPage.js:43]: This should occur in the talon.

File: packages/venia-ui/lib/i18n/en_US.json
Comment (sirugh) [inline packages/venia-ui/lib/i18n/en_US.json:42]: Ah, yes, here I would use id format like the above `component.thing` rather than a string.

**Review by sirugh — CHANGES_REQUESTED** [review]

File: packages/peregrine/lib/talons/RootComponents/Category/useCategoryContent.js
Comment (eug123) [reply packages/peregrine/lib/talons/RootComponents/Category/useCategoryContent.js]: Done, 
Thank you

File: packages/peregrine/lib/talons/CategoryTree/useCategoryTree.js
Comment (eug123) [reply packages/peregrine/lib/talons/CategoryTree/useCategoryTree.js]: Fixed, 
Thank you

File: packages/peregrine/lib/talons/CategoryTree/useCategoryTree.js
Comment (eug123) [reply packages/peregrine/lib/talons/CategoryTree/useCategoryTree.js]: Changed,
Thank you

File: packages/peregrine/lib/talons/Header/useStoreSwitcher.js
Comment (eug123) [reply packages/peregrine/lib/talons/Header/useStoreSwitcher.js:71]: Comment added,
Thank you

File: packages/venia-ui/lib/components/MagentoRoute/magentoRoute.js
Comment (eug123) [reply packages/venia-ui/lib/components/MagentoRoute/magentoRoute.js]: Done

File: packages/venia-ui/lib/components/Navigation/navigation.js
Comment (eug123) [reply packages/venia-ui/lib/components/Navigation/navigation.js]: This allows not having two switchers in the same time in header and navigation panel.

File: packages/venia-ui/lib/components/Header/header.css
Comment (eug123) [reply packages/venia-ui/lib/components/Header/header.css]: Change to be header local variable and value is set in header component and depends on header inner layout.

File: packages/venia-ui/lib/components/SearchPage/searchPage.js
Comment (eug123) [reply packages/venia-ui/lib/components/SearchPage/searchPage.js]: Done

File: packages/venia-ui/lib/i18n/en_US.json
Comment (eug123) [reply packages/venia-ui/lib/i18n/en_US.json]: Done,
Thank you

Comment (sirugh) [reply packages/venia-ui/lib/i18n/en_US.json]: Thanks for updating it - seems like we're still going to be figuring this out for a while, but I don't think you need to change it again in your PR.

File: packages/venia-ui/lib/components/Header/header.css
Comment (sirugh) [reply packages/venia-ui/lib/components/Header/header.css]: Ah I see. You did it like this:

```
const style = {
        ...(switchers && { '--header-height': '7.5rem' })
    };

    return (
        <header className={rootClass} style={style}>
        ...
```

File: packages/venia-ui/lib/components/Navigation/navigation.js
Comment (sirugh) [reply packages/venia-ui/lib/components/Navigation/navigation.js]: Is having two switchers rendered a concern?

File: packages/venia-ui/lib/components/Header/switcherItem.js
Comment (sirugh) [inline packages/venia-ui/lib/components/Header/switcherItem.js:20]: Rather than change the click handler you can just set `disabled={active}`

File: packages/venia-ui/lib/components/Navigation/navigation.js
Comment (sirugh) [inline packages/venia-ui/lib/components/Navigation/navigation.js:58]: Will `StoreSwitcher` ever be falsy? I don't think you need this conditional.

File: packages/peregrine/lib/talons/MagentoRoute/useMagentoRoute.js
Comment (sirugh) [inline packages/peregrine/lib/talons/MagentoRoute/useMagentoRoute.js:72]: Can move this `store` truthiness check into the `shouldFetch` utility.

File: packages/peregrine/lib/talons/Header/useStoreSwitcher.js
Comment (sirugh) [inline packages/peregrine/lib/talons/Header/useStoreSwitcher.js:24]: This should be initialized outside of the talon.

**Review by sirugh — COMMENTED** [review]
More feedback :) I have concerns about the mobile views but this is mostly ready to go IMO.

File: packages/venia-ui/lib/components/Navigation/navigation.js
Comment (eug123) [reply packages/venia-ui/lib/components/Navigation/navigation.js:58]: It may be when there are only one store view so nothing to switch between, currency switcher may go to `<div class="switchers"/>` too, and they are both independently can be available/unavailable

Comment (sirugh) [reply packages/venia-ui/lib/components/Navigation/navigation.js:58]: But the export itself will always return the function component. This conditional rendering should probably occur within the switcher component itself.

File: packages/peregrine/lib/talons/Header/useStoreSwitcher.js
Comment (eug123) [reply packages/peregrine/lib/talons/Header/useStoreSwitcher.js]: Done, thank you

File: packages/peregrine/lib/talons/MagentoRoute/useMagentoRoute.js
Comment (eug123) [reply packages/peregrine/lib/talons/MagentoRoute/useMagentoRoute.js]: Done

File: packages/venia-ui/lib/components/Header/switcherItem.js
Comment (eug123) [reply packages/venia-ui/lib/components/Header/switcherItem.js]: Changed this.

Thank you

File: packages/venia-ui/lib/components/Navigation/navigation.js
Comment (eug123) [reply packages/venia-ui/lib/components/Navigation/navigation.js]: You are right, I need this to have control over `--header-height` custom property, but it's never worked. I have changed to `useRef` to get switchers height.

File: packages/venia-ui/lib/components/Navigation/navigation.js
Comment (eug123) [reply packages/venia-ui/lib/components/Navigation/navigation.js]: Resolved this on `header` and `navigation` components level with media queries for `.switchers` container

File: packages/venia-ui/lib/queries/getAvailableStoresConfigData.graphql
Comment (sirugh) [inline packages/venia-ui/lib/queries/getAvailableStoresConfigData.graphql:1]: Instead of generic `queries` we are moving to colocating queries next to their components that utilize them.

So this would become:

```
import { GET_CONFIG_DATA_QUERY } from './storeSwitcher.gql.js';
```

And you would modify it to use `gql` and export the query.

I'd like to see this new query rewritten in this manner, but the other queries you are just modifying in this change set I leave to your discretion.

**Review by sirugh — CHANGES_REQUESTED** [review]
I've been through this PR a lot with you and I think I've given all the feedback I have to give. I would like to see someone else like @jimbo review it before we merge.

There are also some small changes I'd like before approving, and there is also a comment from brabender [here](https://github.com/magento/pwa-studio/pull/2696#issuecomment-694329608) about some issues he had with this and dave's PR combined. Please work with Dave to resolve that.

File: packages/venia-ui/lib/queries/getAvailableStoresConfigData.graphql
Comment (eug123) [reply packages/venia-ui/lib/queries/getAvailableStoresConfigData.graphql]: Moved query next to storeSwitcher component.
Regarding this [issue](https://github.com/magento/pwa-studio/pull/2696#issuecomment-694329608), I've tried to merge Dave's branch into mine before, there were some conflicts, but at the first look it seems to work fine, I'm starting more thorough testing both branches merged now, and will push it, when Dave's PR will be merged.

File: packages/venia-concept/.graphqlconfig
Comment (sirugh) [inline packages/venia-concept/.graphqlconfig:18]: What is the reason for ignoring the store switcher queries?

File: packages/venia-ui/lib/queries/getStoreConfigData.graphql
Comment (sirugh) [inline packages/venia-ui/lib/queries/getStoreConfigData.graphql:7]: The category and searchPage components should each have their own queries for `storeConfig` data and it should be pared down to _just_ the data those components need. Infact it looks like they both only use `grid_per_page` so you should be able to create a much smaller query next to each:

```gql
const getPageSize = gql`
  query getPageSize {
    storeConfig {
      grid_per_page
    }
  }
}
```

File: packages/venia-ui/lib/queries/getStoreConfigData.graphql
Comment (sirugh) [inline packages/venia-ui/lib/queries/getStoreConfigData.graphql]: You can remove the other properties from this query that aren't used.

File: packages/venia-ui/lib/queries/getStoreConfigData.graphql
Comment (sirugh) [reply packages/venia-ui/lib/queries/getStoreConfigData.graphql:7]: And you can remove `grid_per_page` from this query.

**Review by sirugh — CHANGES_REQUESTED** [review]
Can we get an approval from someone else besides me? This is a pretty big change and I want to get more than just my eyes on the whole thing.

File: packages/venia-concept/.graphqlconfig
Comment (eug123) [reply packages/venia-concept/.graphqlconfig:18]: It contains query only available since Magento 2.4.1

File: packages/venia-ui/lib/queries/getStoreConfigData.graphql
Comment (eug123) [reply packages/venia-ui/lib/queries/getStoreConfigData.graphql]: Moved queries for `storeConfig` to respective components

File: packages/venia-concept/.graphqlconfig
Comment (sirugh) [reply packages/venia-concept/.graphqlconfig:18]: Hmm is there a way we can indicate a comment in this page or do we just have to remember to remove these when 2.4.1 is released?

File: packages/venia-ui/lib/components/App/localeProvider.js
Comment (sirugh) [reply packages/venia-ui/lib/components/App/localeProvider.js:64]: I know it was causing a warning but I'd like to try and figure out why before we just merge this line. If you change it to 

```js
if (loading && !data) {
   debugger;
  return fullPageLoadingIndicator
}
```

does it pause after you change the locales?

If so, it means that for some reason data is getting deleted even after you refresh the page,

You might also try changing this to `if (!messages)` which should be the same thing.

File: packages/venia-ui/lib/queries/getStoreConfigData.graphql
Comment (eug123) [reply packages/venia-ui/lib/queries/getStoreConfigData.graphql]: Done

File: packages/venia-ui/lib/components/App/localeProvider.js
Comment (eug123) [reply packages/venia-ui/lib/components/App/localeProvider.js:64]: It doesn't step into condition `if (loading && !data)` and stop on debugger when I switch locales. Since we already have `data` cached `storeConfig` for all stores in `storeSwitcher`

It does stop on `if (loading && !messages)` and just `if (!messages)`, but still got warning.

Leaving just `if (loading)` works without warning

Comment (sirugh) [reply packages/venia-ui/lib/components/App/localeProvider.js:64]: OK. That's fine. No need to debug the deeper issue with the warning right now.

File: packages/venia-concept/.graphqlconfig
Comment (eug123) [reply packages/venia-concept/.graphqlconfig:18]: Add comment to `storeSwitcher.gql.js`

**Review by sirugh — DISMISSED** [review]

File: packages/peregrine/lib/talons/Header/useStoreSwitcher.js
Comment (jimbo) [inline packages/peregrine/lib/talons/Header/useStoreSwitcher.js:14]: If you're going to reduce an iterable (`rawAvailableStoresData` here) to a keyed object, you should consider whether the consumer will

- **know** or **not know** the keys
- **look up** values or **iterate over** entries

In this case, the consumer (component) will **not know** the keys and will **iterate over** entries to create elements, so it'd be better to reduce to a `Map` than a plain object.

```es6
const getAvailableOptions = (config, stores) => {
  const { code: configCode } = config

  const options = stores.reduce(
    (map, store) => {
      const {
        code,
        default_display_currency_code: currency,
        locale,
        store_name: storeName
      } = store

      const isCurrent = code === configCode
      const option = { currency, isCurrent, locale, storeName }

      return map.set(code, option)
    },
    new Map()
  )
}
```

Also, the option object here should probably use `isCurrent` instead of `is_current`, to match the casing of `storeName`.

File: packages/peregrine/lib/talons/Header/useStoreSwitcher.js
Comment (jimbo) [inline packages/peregrine/lib/talons/Header/useStoreSwitcher.js:71]:
```suggestion
        // Change store view code and currency to be used in Apollo link request headers
```

**Review by jimbo — CHANGES_REQUESTED** [review]
It would be nice if we could just use the mobile store switcher UI, where the switcher is in the main drawer, at all resolutions. I think it's much better than the header toolbar, comes with way fewer complications, and works perfectly in the drawer.

File: packages/peregrine/lib/talons/Header/useStoreSwitcher.js
Comment (eug123) [reply packages/peregrine/lib/talons/Header/useStoreSwitcher.js]: Done, thank you

File: packages/peregrine/lib/talons/Header/useStoreSwitcher.js
Comment (eug123) [reply packages/peregrine/lib/talons/Header/useStoreSwitcher.js]: Corrected

File: packages/peregrine/lib/talons/CategoryTree/useCategoryTree.js
Comment (sirugh) [reply packages/peregrine/lib/talons/CategoryTree/useCategoryTree.js:46]: The catalog data is still necessary even after this change as it controls the rendering near the top of this tree in `useNavigation`. All that @eug123 did here is change where the category tree (this component) gets the data - instead of using the prop data it just uses the query data that it was making anyways. Functionally identical without the round trip of updating the store and then re-rendering when the `catalog` prop makes its way back here.

I do think we should try to disentangle some of the catalog reducer logic from the navigation drawer but I think that scope is beyond this PR.

Comment (eug123) [reply packages/peregrine/lib/talons/CategoryTree/useCategoryTree.js:46]: Other components still use `categories` from Redux. Seems removing `updateCategories()` from here, will require rebuild all of them to use Apollo cache, and eventually deprecate `useCatalogContext`

**Review by jimbo — DISMISSED** [review]

**Comment by fooman** [comment]
One thing which I am not sure is handled yet is that different store views can very likely have different root_category_ids. When switching this should also update.

**Comment by dpatil-magento** [comment]
@eug123 Please check below observations - 
1.Safari Icon and text are too close.
image
2.Even though #STORE_VIEW_CODE=fr is commented, store switcher still displays when backend have multiple store views. Expected?  I was thinking store switcher would only be shown when  STORE_VIEW_CODE is enabled in .env.
3. Respective store categories are not maintained. i.e. Default Store specific categories displayed for French store also.

**Comment by dpatil-magento** [comment]
1 is fixed, 2 as designed, 3 backend data fixed.

QA Approved.

**Comment by eug123** [comment]
@fooman, updating root category is not handled by this PR, which covers basic scenario when all store views are in the one store group. Since root_category is store group level setting, it will be addressed in the upcoming Multi-Store Support epic.

File: packages/venia-ui/lib/RootComponents/Category/category.js
Comment (sirugh) [inline packages/venia-ui/lib/RootComponents/Category/category.js:47]: This is a good change but I would move this `if` block above the error case block and then remove the `&& !loading`.

File: packages/peregrine/lib/talons/RootComponents/Category/useCategory.js
Comment (sirugh) [inline packages/peregrine/lib/talons/RootComponents/Category/useCategory.js:203]: From the perspective of the UI component `loading` should be true if and only if we have do not have data and there are in flight requests.

You can simplify this:

```
const loading = !data && categoryLoading;
```

Reference [this example](https://github.com/magento/pwa-studio/blob/develop/packages/peregrine/lib/talons/OrderHistoryPage/useOrderHistoryPage.js#L29-L49) to see how we use data and loading states to determine when to render the full page loading indicator versus the background loading indicator.

**Review by sirugh — CHANGES_REQUESTED** [review]
Hopefully last round of feedback!

File: packages/venia-ui/lib/policies/index.js
Comment (eug123) [reply packages/venia-ui/lib/policies/index.js]: Removed, thank you

File: packages/venia-ui/lib/components/MagentoRoute/magentoRoute.js
Comment (eug123) [reply packages/venia-ui/lib/components/MagentoRoute/magentoRoute.js:19]: There are more queries added with this PR, probably it is better to align them with new [pattern](https://github.com/magento/pwa-studio/pull/2712) in one separate PR

File: packages/peregrine/lib/talons/SearchPage/useSearchPage.js
Comment (eug123) [reply packages/peregrine/lib/talons/SearchPage/useSearchPage.js:141]: Done

File: packages/peregrine/lib/talons/RootComponents/Category/useCategory.js
Comment (eug123) [reply packages/peregrine/lib/talons/RootComponents/Category/useCategory.js:137]: Done

File: packages/peregrine/lib/talons/RootComponents/Category/useCategory.js
Comment (eug123) [reply packages/peregrine/lib/talons/RootComponents/Category/useCategory.js:203]: It is done similar to [useSearchPage](https://github.com/magento/pwa-studio/blob/develop/packages/peregrine/lib/talons/SearchPage/useSearchPage.js#L198), when only categoryLoading is involved, `noProductsFound` component flashes for a moment.

File: packages/venia-ui/lib/components/App/localeProvider.js
Comment (eug123) [reply packages/venia-ui/lib/components/App/localeProvider.js:46]: Fixed blocking error [here](https://github.com/magento/pwa-studio/pull/2686/files?file-filters%5B%5D=.js#diff-837aa57217caa0b1c16e1ca9e5461009R94)

Comment (sirugh) [reply packages/venia-ui/lib/components/App/localeProvider.js:46]: Awesome, thanks.

File: packages/peregrine/lib/talons/RootComponents/Category/useCategory.js
Comment (sirugh) [reply packages/peregrine/lib/talons/RootComponents/Category/useCategory.js:203]: Ok, please leave a comment indicating it at least, since it is an outlier.

Comment (eug123) [reply packages/peregrine/lib/talons/RootComponents/Category/useCategory.js:204]: Done

**Review by sirugh — APPROVED** [review]
APPROVED! Hopefully for the last time :D

**Comment by dpatil-magento** [comment]
QA re-approved.

**Pull request merged**
